src/main/kotlin/org/rust/ide/inspections/lints/RsPathUsageAnalysis.kt

```diff
@@ -0,0 +1,108 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections.lints
+
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.CachedValue
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.infer.ResolvedPath
+import org.rust.lang.core.types.inference
+import org.rust.openapiext.processElementsWithMacros
+
+data class PathUsageMap(
+    val pathUsages: Map<String, Set<RsElement>>,
+    val traitUsages: Set<RsTraitItem>
+)
+
+private val PATH_USAGE_KEY: Key<CachedValue<PathUsageMap>> = Key.create("PATH_USAGE_KEY")
+
+val RsItemsOwner.pathUsage: PathUsageMap
+    get() = CachedValuesManager.getCachedValue(this, PATH_USAGE_KEY) {
+        val usages = calculatePathUsages(this)
+        CachedValueProvider.Result.create(usages, PsiModificationTracker.MODIFICATION_COUNT)
+    }
+
+private fun calculatePathUsages(owner: RsItemsOwner): PathUsageMap {
+    val directUsages = hashMapOf<String, MutableSet<RsElement>>()
+    val traitUsages = hashSetOf<RsTraitItem>()
+
+    for (child in owner.children) {
+        processElementsWithMacros(child) { element ->
```

Comment:
We should also process macro paths here

```rust
use lazy_static::lazy_static;

lazy_static! {
    static ref EXAMPLE: u8 = 42;
}
```

Reply:
I'm not sure how to do this properly because of the unresolved macro problems mentioned in https://github.com/intellij-rust/intellij-rust/pull/6392#issuecomment-853369494.

Reply:
Let's then ignore imports which resolve to macros in this PR, and I will think how to best handle macros